accepted/2021/System.Diagnostics/Metrics-Design.md

```diff
@@ -0,0 +1,511 @@
+                            params KeyValuePair<string, object?>[] tags) { throw null; }
+    }
+
+    /// <summary>
+    /// UpDownCounter is a non-observable Instrument that supports increments and decrements.
+    /// e.g. Number of items in a queue.
+    /// </summary>
+    public class UpDownCounter<T> : Instrument<T> where T : unmanaged
+    {
+        public void Add(T measurement) { throw null; }
+        public void Add(T measurement,
+                            KeyValuePair<string, object?> tag1) { throw null; }
+        public void Add(T measurement,
+                            KeyValuePair<string, object?> tag1,
+                            KeyValuePair<string, object?> tag2) { throw null; }
+        public void Add(T measurement,
+                            KeyValuePair<string, object?> tag1,
+                            KeyValuePair<string, object?> tag2,
+                            KeyValuePair<string, object?> tag3) { throw null; }
+        public void Add(T measurement,
+                            ReadOnlySpan<<string, object?>> tags) { throw null; }
+        public void Add(T measurement,
+                            params KeyValuePair<string, object?>[] tags) { throw null; }
+    }
+
+    /// <summary>
+    /// The histogram is a non-observable Instrument that can be used to report arbitrary values
+    /// that are likely to be statistically meaningful. It is intended for statistics such
+    /// e.g. the request duration.
+    /// </summary>
+    public class Histogram<T> : Instrument<T> where T : unmanaged
+    {
+        public void Record(T measurement) { throw null; }
+        public void Record(
+                        T measurement,
+                        KeyValuePair<string, object?> tag1) { throw null; }
+        public void Record(
+                        T measurement,
+                        KeyValuePair<string, object?> tag1,
+                        KeyValuePair<string, object?> tag2) { throw null; }
+        public void Record(
+                        T measurement,
+                        KeyValuePair<string, object?> tag1,
+                        KeyValuePair<string, object?> tag2,
+                        KeyValuePair<string, object?> tag3) { throw null; }
+        public void Record(T measurement,
+                            ReadOnlySpan<<string, object?>> tags) { throw null; }
+        public void Record(
+                        T measurement,
+                        params KeyValuePair<string, object?>[] tags) { throw null; }
+    }
+}
+
+```
+
+### Observable Instruments
+
+```csharp
+namespace System.Diagnostics.Metrics
+{
+    /// <summary>
+    /// ObservableCounter is an observable Instrument that reports monotonically increasing value(s)
+    /// when the instrument is being observed.
+    /// e.g. CPU time (for different processes, threads, user mode or kernel mode).
+    /// </summary>
+    public class ObservableCounter<T> : ObservableInstrument<T> where T : unmanaged
```

Comment:
Should we seal all these derived instrument implementations? We haven't planned for any scenarios where users would directly extend them and I assume we could always unseal them in the future if a new scenario emerged.

Reply:
agree. it make sense to seal them.

Reply:
@noahfalk 
The question now about the listener. can you think in any scenario someone would need to extend it? if not, then we can seal it too.

Reply:
I don't see any reason the listener needs to be extendable and I'd be fine to seal it for now. At best I could see scenarios where it is mildly more convenient for the user to extend the listener rather than aggregating it, but I don't see a reason why we need to prioritize it up-front.

Reply:
I sealed the listener too.